cpp/FbxCliWrapper/FbxCliWriter.cpp

```diff
@@ -0,0 +1,188 @@
+#include "FbxCliWriter.h"
+#include <assert.h>
+
+
+namespace FbxClrWrapper
+{
+	int FbxCliWriter::SaveFBX(String^ FileName)
+	{
+		if (!TransformDataFromCLI())
+		{
+			return -1;
+		}
+
+		auto fileName = (const char*)(void*)System::Runtime::InteropServices::Marshal::StringToHGlobalAnsi(FileName);
+		bool lResult = SaveScene(fileName);
+		if (lResult)
+		{
+			return 0;
+		}
+
+		return -1;
+	}
+
+	bool FbxCliWriter::SaveScene(const char* pFilename)
+	{
+		// Create an IOSettings object.
+		FbxIOSettings* ios = FbxIOSettings::Create(mSdkManager, IOSROOT);
+		mSdkManager->SetIOSettings(ios);
+
+		// ... Configure the FbxIOSettings object here ...
+
+		ios->SetBoolProp(IMP_FBX_GLOBAL_SETTINGS, true);
+
+		// Create an exporter.
+		FbxExporter* lExporter = FbxExporter::Create(mSdkManager, "");
+
+		// Initialize the exporter.
+		bool lExportStatus = lExporter->Initialize(pFilename, -1, mSdkManager->GetIOSettings());
+
+
+		if (!lExportStatus) 
+		{
+			printf("Call to FbxExporter::Initialize() failed.\n");
+			printf("Error returned: %s\n\n", lExporter->GetStatus().GetErrorString());
+			return false;
+		}
+
+		// Create a new scene so it can be populated by the CLI Data.
+		mScene = FbxScene::Create(mSdkManager, "myScene");
+
+		FbxAxisSystem newAxisSystem(FbxAxisSystem::EUpVector::eYAxis,
+			FbxAxisSystem::EFrontVector::eParityOdd,
+			FbxAxisSystem::ECoordSystem::eRightHanded);
+
+		mScene->GetGlobalSettings().SetAxisSystem(newAxisSystem);
+
+		ExportNodes();
+
+		// Export the scene to the file.
+		bool result = lExporter->Export(mScene);
+
+		if (!result)
+		{
+			auto status = lExporter->GetStatus();
+			printf("Error: %s", status.GetErrorString());
+		}
+
+		// Destroy the exporter.
+		lExporter->Destroy();
+
+		return result;
+	}
+
+	void FbxCliWriter::ExportNodes()
+	{
+		// Create all the FBX meshes and nodes
+		std::vector<FbxMesh*> meshList;
+		CreateFBXMeshList(meshList);
+		CreateFBXNodes(meshList);
+	}
+
+	void FbxCliWriter::CreateFBXNodes(std::vector<FbxMesh*>& MeshList)
+	{
+		// Create all the fbx nodes required
+		// Skip the IScene's root node
+		FbxNode* lRootNode = mScene->GetRootNode();
+		std::vector<FbxNode*> nodeList;
+		for (int nodeIndex = 1; nodeIndex < (int)mSceneData->mNodeNameList.size(); nodeIndex++)
+		{
+			auto nodeName = mSceneData->mNodeNameList[nodeIndex];
+
+			FbxNode* newNode = FbxNode::Create(mScene, nodeName.c_str());
+			nodeList.push_back(newNode);
+		}
+
+		// use the parent index list to insert children into correct nodes
+		for (int nodeIndex = 0; nodeIndex < mSceneData->mNodeParentList.size(); nodeIndex++)
+		{
+			int32_t meshIndex = mSceneData->mNodeMeshIndexList[nodeIndex];
+			int32_t nodeParentIndex = mSceneData->mNodeParentList[nodeIndex];
+			FbxMatrix nodeLocalTransform;
+			auto node = nodeIndex == 0 ? lRootNode : nodeList[nodeIndex - 1];
+
+			if (nodeParentIndex == -1)
+			{
+				nodeLocalTransform = *reinterpret_cast<FbxMatrix*>(&mSceneData->mNodeTransformList[nodeIndex]);

```

Comment:
are these all really `reinterpret_cast` casts? I would have expected a `static_cast` to do the trick.